include/deal.II/dofs/dof_tools.h

```diff
@@ -1153,12 +1153,12 @@ namespace DoFTools
    *
    * @author Matthias Maier, 2012 - 2015
    */
-  template <typename FaceIterator>
+  template <typename FaceIterator, typename number>
   void
   make_periodicity_constraints(
     const FaceIterator &                         face_1,
     const typename identity<FaceIterator>::type &face_2,
-    AffineConstraints<double> &                  constraints,
+    AffineConstraints<number> &                  constraints,
```

Comment:
Don't you want to use `FullMatrix<number>` instead of `FullMatrix<double>` below?

Reply:
As this is an item on my (ever growing) todo list: I would very much like to see full complex number support in `make_periodicity_constraints` - including a phase shift over periodic boundaries. The hard part for this (and why this was not in my initial patch for `AffineConstraints` is the fact that we have to transform all the interpolation matrices and rotation matrices that are getting generated recursively also to complex valued entities.

@dangars Would you be open to try to tackle this as part of this PR?

Reply:
@masterleinad That is right for the particular application I'm working on right now, this patch is enough because I just need to initialize an `AffineConstraints<std::complex<double>>` object.

@tamiko In the near future I want to compute the bandstructure of a phononic and phononic device. For that I need to impose a phase shift over periodic boundaries. I guess that is the application that you have in mind. Yes, let's do the phase shift support in this PR!

On my todo list there is a tutorial on how to calculate the bandgap using a periodic shift. I'll start to work on that after this PR is finished.

@tamiko I started to update the rotation matrix and I see all the dependencies. As @masterleinad suggested I'll update as well `matrix` in `DoFTools::make_hanging_node_constraints` and `matrix` and `offset` in `GridTools::collect_periodic_faces` to be complex-valued.

Can you add the WIP label? I'll let you know when it is ready for review :)